perfkitbenchmarker/pkb.py

```diff
@@ -195,6 +195,51 @@
 events.initialization_complete.connect(traces.RegisterAll)
 
 
+def _CreateBenchmarkRunList():
+  """Create a list of configs and states for each benchmark run to be scheduled.
+
+  Returns:
+    list of (args, run_status_list) pairs. Contains one pair per benchmark run
+    to be scheduled, including multiple pairs for benchmarks that will be run
+    multiple times. args is a tuple of the first five arguments to pass to
+    RunBenchmark, and run_status_list is a list of strings in the order of
+    [benchmark_name, benchmark_uid, benchmark_status].
+  """
+  result = []
+  benchmark_tuple_list = benchmark_sets.GetBenchmarksFromFlags()
+  total_benchmarks = len(benchmark_tuple_list)
+  benchmark_counts = collections.defaultdict(itertools.count)
+  for i, benchmark_tuple in enumerate(benchmark_tuple_list):
+    # Construct benchmark config object.
+    benchmark_module, user_config = benchmark_tuple
+    name = benchmark_module.BENCHMARK_NAME
+    config_dict = benchmark_module.GetConfig(user_config)
+    config_spec_class = getattr(
+        benchmark_module, 'BENCHMARK_CONFIG_SPEC_CLASS',
+        benchmark_config_spec.BenchmarkConfigSpec)
+    config = config_spec_class(name, flag_values=FLAGS, **config_dict)
+
+    # Assign a unique ID to each benchmark run. This differs even between two
+    # runs of the same benchmark within a single PKB run.
+    uid = name + str(benchmark_counts[name].next())
+
+    # Optional step to check flag values and verify files exist.
+    check_prereqs = getattr(benchmark_module, 'CheckPrerequisites', None)
+    if check_prereqs:
+      try:
+        with config.RedirectFlags(FLAGS):
+          check_prereqs()
```

Comment:
I think that's generally reasonable, but there may be a case where it's less desirable. Currently, we allow re-running stages with mixed levels of support depending on the benchmark (e.g. provision-prepare-run-run-run-cleanup-teardown, or provision-prepare-run-cleanup-prepare-run-cleanup-teardown). This allows changing a few benchmark parameters and rerunning the experiment on the same resources. The problem is that there are some inputs (FLAGS) that control the properties of the provisioned resources, which don't make sense to change between running stages, and there are some inputs that control other benchmark behavior. To my knowledge, the latter only applies to the following benchmarks:
- cassandra_stress
- cloudsuite_websearch
- speccpu2006
- specsfs2014
- ycsb benchmarks
which all involve copying config files to the VMs (prepare stage). However, if we're running in stages and repeating the prepare stage, there's less concern about early failure (since the resources have already been provisioned), but it means a failure may arise in the middle of Prepare when trying to copy the files rather than before Prepare, in CheckPrerequisites.

So I think what you're suggesting is reasonable, but it may merit more thought as a separate issue along with the whole repeating stages discussion. Currently, aside from speccpu2006 peeking into a tar file, there aren't any checks that take a lot of time, so there wouldn't be a big performance drawback to letting the call happen in all stages by default.